File: app/app/src/main/java/com/example/hakonsreader/views/util/MenuClickHandler.java
package com.example.hakonsreader.views.util;

import android.app.AlertDialog;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.widget.PopupMenu;

import androidx.annotation.IdRes;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import com.example.hakonsreader.App;
import com.example.hakonsreader.R;
import com.example.hakonsreader.api.enums.PostTimeSort;
import com.example.hakonsreader.dialogadapters.OAuthScopeAdapter;
import com.example.hakonsreader.interfaces.SortableWithTime;
import com.example.hakonsreader.misc.TokenManager;

import java.util.ArrayList;
import java.util.Arrays;

/**
 * Class containing click handlers for menus
 */
public class MenuClickHandler {
    private static final String TAG = "MenuClickHandler";

    private MenuClickHandler() { }

    /**
     * Click handler for the profile "more" (kebab) button. Shows a popup menu customized
     * based on if it should be shown for a logged in user or not
     *
     * @param view The view clicked (where the menu will be attached)
     * @param loggedInUser True if the button is clicked for a logged in user
     */
    public static void showPopupForProfile(View view, boolean loggedInUser) {
        if (loggedInUser) {
            showPopupForProfileForLoggedInUser(view);
        } else {
            showPopupForProfileForNonLoggedInUser(view);
        }
    }

    /**
     * Shows the popup menu for profiles for logged in users
     *
     * @param view The view clicked (where the menu will be attached)
     */
    public static void showPopupForProfileForLoggedInUser(View view) {
        PopupMenu menu = new PopupMenu(view.getContext(), view);
        menu.inflate(R.menu.profile_menu_logged_in_user);

        boolean isPrivatelyBrowsing = App.Companion.get().getApi().isPrivatelyBrowsing();
        if (isPrivatelyBrowsing) {
            MenuItem savedItem = menu.getMenu().findItem(R.id.menuBrowsePrivately);
            savedItem.setTitle(view.getContext().getString(R.string.menuPrivateBrowsingDisable));
        }

        menu.setOnMenuItemClickListener(item -> {
            int itemId = item.getItemId();

            if (itemId == R.id.menuLogOut) {
                App.Companion.get().logOut();
                return true;
            } else if (itemId == R.id.menuBrowsePrivately) {
                App.Companion.get().enablePrivateBrowsing(!isPrivatelyBrowsing);
                return true;
            } else if (itemId == R.id.menuApplicationPrivileges) {
                showApplicationPrivileges(view.getContext(), view.getParent());
                return true;
            }

            return false;
        });

        menu.show();
    }

    /**
     * Shows a popup of the applications OAuth privileges
     */
    private static void showApplicationPrivileges(Context context, ViewParent parent) {
        // TODO if scopes have been added to the application that isn't in the stored token, show which are missing as well
        ArrayList<String> scopes = new ArrayList<>(Arrays.asList(TokenManager.getToken().getScopesAsArray()));
        OAuthScopeAdapter adapter = new OAuthScopeAdapter(context, R.layout.list_item_oauth_explanation, scopes);

        View title = LayoutInflater.from(context).inflate(R.layout.dialog_title_oauth_explanation, (ViewGroup) parent, false);

        new AlertDialog.Builder(context)
                .setCustomTitle(title)
                .setAdapter(adapter, null)
                .show();
    }

    /**
     * Shows the popup menu for profiles for non-logged in users
     *
     * @param view The view clicked (where the menu will be attached)
     */
    public static void showPopupForProfileForNonLoggedInUser(View view) {

    }

    /**
     * Shows a popup menu to allow a list to change how it should be sorted. The menu shown here
     * includes time sorts for sorts such as top and controversial
     *
     * @param view The view clicked. If this view is not a child of a fragment or activity implementing
     *             {@link SortableWithTime} nothing is done
     */
    public static void showPopupSortWithTime(View view) {
        Fragment f = FragmentManager.findFragment(view);
        Context context = view.getContext();

        SortableWithTime sortable;

        if (f instanceof SortableWithTime) {
            sortable = (SortableWithTime) f;
        } else if (context instanceof SortableWithTime) {
            sortable = (SortableWithTime) context;
        } else {
            return;
        }
        // TODO the menu should show what is the currently selected sort (probably have to adjust the interface for a getter method)

        PopupMenu menu = new PopupMenu(context, view);
        menu.inflate(R.menu.sort_menu_with_time);

        menu.setOnMenuItemClickListener(item -> {
            int itemId = item.getItemId();

            if (itemId == R.id.sortNew) {
                sortable.newSort();
                return true;
            } else if (itemId == R.id.sortHot) {
                sortable.hot();
                return true;
            } else if (itemId == R.id.sortTop) {
                // If top/controversial inflate new menu with sort_menu_of_time
                MenuItem subMenu = menu.getMenu().findItem(R.id.sortTop);
                menu.getMenuInflater().inflate(R.menu.sort_times, subMenu.getSubMenu());
                menu.setOnMenuItemClickListener(subItem -> {
                    int subItemId = subItem.getItemId();
                    sortable.top(getTimeSortFromId(subItemId));
                    return true;
                });
                return true;
            } else if (itemId == R.id.sortControversial) {
                MenuItem subMenu = menu.getMenu().findItem(R.id.sortControversial);
                menu.getMenuInflater().inflate(R.menu.sort_times, subMenu.getSubMenu());
                menu.setOnMenuItemClickListener(subItem -> {
                    int subItemId = subItem.getItemId();
                    sortable.controversial(getTimeSortFromId(subItemId));
                    return true;
                });
                return true;
            }

            return false;
        });

        menu.show();
    }


    /**
     * Retrieves the correct {@link PostTimeSort} based on an ID res
     *
     * @param id The ID to retrieve a sort for (the IDs from {@code sort_times.xml}
     * @return The corresponding {@link PostTimeSort}. Default if no match is {@link PostTimeSort#ALL_TIME}
     */
    private static PostTimeSort getTimeSortFromId(@IdRes int id) {
        PostTimeSort timeSort;

        if (id == R.id.sortNow) {
            timeSort = PostTimeSort.HOUR;
        } else if (id == R.id.sortToday) {
            timeSort = PostTimeSort.DAY;
        } else if (id == R.id.sortWeek) {
            timeSort = PostTimeSort.WEEK;
        } else if(id == R.id.sortMonth) {
            timeSort = PostTimeSort.MONTH;
        } else if (id == R.id.sortYear) {
            timeSort = PostTimeSort.YEAR;
        } else {
            timeSort = PostTimeSort.ALL_TIME;
        }

        return timeSort;
    }
}

File: app/app/src/main/java/com/example/hakonsreader/views/util/MenuClickHandler.kt
package com.example.hakonsreader.views.util

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.content.res.Resources
import android.view.*
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.hakonsreader.R
import com.example.hakonsreader.api.RedditApi
import com.example.hakonsreader.api.enums.PostTimeSort
import com.example.hakonsreader.api.model.RedditUser
import com.example.hakonsreader.api.persistence.RedditUserInfoDao
import com.example.hakonsreader.databinding.DialogAccountManagementBinding
import com.example.hakonsreader.dialogadapters.OAuthScopeAdapter
import com.example.hakonsreader.fragments.PostsFragment
import com.example.hakonsreader.interfaces.SortableWithTime
import com.example.hakonsreader.misc.*
import com.example.hakonsreader.recyclerviewadapters.AccountsAdapter
import com.example.hakonsreader.states.AppState
import com.example.hakonsreader.states.LoggedInState
import com.github.zawadz88.materialpopupmenu.popupMenu
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.Dispatchers.Main
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.*

/**
 * Shows the popup menu for profiles for logged in users
 *
 * @param view The view clicked (where the menu will be attached)
 * @param user The user represented in the profile popup clicked
 * @param api The API object to use for API calls
 */
fun showPopupForProfile(view: View, user: RedditUser?, api: RedditApi, userInfoDao: RedditUserInfoDao) {
    user ?: return

    if (AppState.getUserInfo()?.userInfo?.id == user.id) {
        showPopupForLoggedInUser(view, api, userInfoDao)
    }
}

private fun showPopupForLoggedInUser(view: View, api: RedditApi, userInfoDao: RedditUserInfoDao) {
    val context = view.context
    val privatelyBrowsing = AppState.loggedInState.value is LoggedInState.PrivatelyBrowsing

    popupMenu {
        style = R.style.Widget_MPM_Menu_Dark_CustomBackground

        section {
            item {
                labelRes = R.string.logOut
                icon = R.drawable.ic_signout
                callback = { AppState.logOut(context) }
            }

            item {
                labelRes = R.string.menuProfileManageAccounts
                icon = R.drawable.ic_baseline_person_24
                callback = { showAccountManagement(view.context, api, userInfoDao) }
            }

            item {
                labelRes = if (privatelyBrowsing) {
                    R.string.menuPrivateBrowsingDisable
                } else {
                    R.string.menuPrivateBrowsingEnable
                }
                icon = R.drawable.ic_incognito
                callback = { AppState.enablePrivateBrowsing(!privatelyBrowsing) }
            }

            // TODO find a better icon for this. The idea for a list icon is "this is a list of
            //  what the app can do". Not sure what a "privileges" icon would even be
            item {
                labelRes = R.string.menuShowApplicationAccessExplanations
                icon = R.drawable.ic_format_list_bulleted_24dp
                callback = { showApplicationPrivileges(context, view.parent) }
            }
        }
    }.show(context, view)
}

/**
 * Shows a Dialog with the list of accounts stored in the application
 *
 * @param context The context used. This should be an activity, otherwise the application cannot recreate
 * itself when a new account is chosen
 */
fun showAccountManagement(context: Context, api: RedditApi, userInfoDao: RedditUserInfoDao) {
    val app = AppState

    Dialog(context).also {
        val binding = DialogAccountManagementBinding.inflate(LayoutInflater.from(context))
        it.setContentView(binding.root)
        it.window?.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)

        binding.addAccount.setOnClickListener {
            startLoginIntent(context)
        }

        binding.accounts.apply {
            layoutManager = LinearLayoutManager(context)
            adapter = AccountsAdapter().apply {
                CoroutineScope(IO).launch {
                    val accs = userInfoDao.getAllUsers() as MutableList
                    withContext(Main) {
                        binding.hasUsers = accs.isNotEmpty()
                        accounts = accs
                    }
                }

                onItemClicked = { userInfoClicked ->
                    val currentId = when (val state = app.loggedInState.value) {
                        is LoggedInState.LoggedIn -> state.userInfo.userId
                        is LoggedInState.PrivatelyBrowsing -> state.userInfo.userId
                        else -> null
                    }

                    if (currentId != userInfoClicked.userId) {
                        // Not sure what we would do if not
                        if (context is AppCompatActivity) {
                            app.switchAccount(userInfoClicked.userId, context)
                        }
                    }
                }

                onRemoveItemClicked = { userInfoClicked ->
                    val currentId = when (val state = app.loggedInState.value) {
                        is LoggedInState.LoggedIn -> state.userInfo.userId
                        is LoggedInState.PrivatelyBrowsing -> state.userInfo.userId
                        else -> null
                    }

                    // Don't remove the item if it's the currently active one
                    if (currentId != userInfoClicked.userId) {
                        val token = TokenManager.getTokenByUserId(userInfoClicked.userId)
                        if (token != null) {
                            removeItem(userInfoClicked)

                            TokenManager.removeTokenByUserId(userInfoClicked.userId)

                            CoroutineScope(IO).launch {
                                userInfoDao.delete(userInfoClicked)
                                api.accessToken().revoke(token)
                            }
                        }
                    }
                }

                onNsfwClicked = { userInfoClicked, nsfwAccount ->
                    val currentId = when (val state = app.loggedInState.value) {
                        is LoggedInState.LoggedIn -> state.userInfo.userId
                        is LoggedInState.PrivatelyBrowsing -> state.userInfo.userId
                        else -> null
                    }

                    // Another account than the active was clicked, update it in the database
                    if (currentId != null && currentId != userInfoClicked.userId) {
                        userInfoClicked.nsfwAccount = nsfwAccount
                        CoroutineScope(IO).launch {
                            userInfoDao.update(userInfoClicked)
                        }
                    } else {
                        // Update the current account
                        CoroutineScope(IO).launch {
                            app.updateUserInfo(nsfwAccount = nsfwAccount)
                        }
                    }
                }
            }
        }

        it.show()
    }
}

/**
 * Shows a popup of the applications OAuth privileges
 */
private fun showApplicationPrivileges(context: Context, parent: ViewParent) {
    // TODO if scopes have been added to the application that isn't in the stored token, show which are missing as well
    val scopes = ArrayList(listOf(*TokenManager.getToken()!!.scopesAsArray))
    val adapter = OAuthScopeAdapter(context, R.layout.list_item_oauth_explanation, scopes)
    val title = LayoutInflater.from(context).inflate(R.layout.dialog_title_oauth_explanation, parent as ViewGroup, false)
    AlertDialog.Builder(context)
            .setCustomTitle(title)
            .setAdapter(adapter, null)
            .show()
}

/**
 * Shows a popup menu to allow a list to change how it should be sorted. The menu shown here
 * includes time sorts for sorts such as top and controversial
 *
 * @param postsFragment The fragment to update the sorting for
 * @param view The view clicked. If this view is not a child of a fragment or activity implementing
 * [SortableWithTime] nothing is done
 */
fun showPopupSortWithTime(postsFragment: PostsFragment, view: View) {
    val context = view.context
    val sortText = getSortText(postsFragment.currentSort(), context)
    val timeSortText = postsFragment.currentTimeSort()?.let { getTimeSortText(it, context) }

    val finalTitle = sortText + if (timeSortText != null) " - $timeSortText" else ""

    // The submenu sizes look weird if they are too small, so set to fixed 40 % of screen width
    val submenuSize = (Resources.getSystem().displayMetrics.widthPixels * 0.4f).toInt()

    popupMenu {
        style = R.style.Widget_MPM_Menu_Dark_CustomBackground

        section {
            title = finalTitle

            item {
                labelRes = R.string.sortNew
                callback = { postsFragment.new() }
            }
            item {
                labelRes = R.string.sortHot
                callback = { postsFragment.hot() }
            }

            item {
                labelRes = R.string.sortTop
                hasNestedItems = true
                callback = {
                    // Show menu
                    popupMenu {
                        style = R.style.Widget_MPM_Menu_Dark_CustomBackground
                        fixedContentWidthInPx = submenuSize

                        section {
                            title = context.getString(R.string.sortTop)

                            item {
                                labelRes = R.string.sortNow
                                callback = { postsFragment.top(PostTimeSort.HOUR) }
                            }
                            item {
                                labelRes = R.string.sortToday
                                callback = { postsFragment.top(PostTimeSort.DAY) }
                            }
                            item {
                                labelRes = R.string.sortWeek
                                callback = { postsFragment.top(PostTimeSort.WEEK) }
                            }
                            item {
                                labelRes = R.string.sortMonth
                                callback = { postsFragment.top(PostTimeSort.MONTH) }
                            }
                            item {
                                labelRes = R.string.sortYear
                                callback = { postsFragment.top(PostTimeSort.YEAR) }
                            }
                            item {
                                labelRes = R.string.sortAllTime
                                callback = { postsFragment.top(PostTimeSort.ALL_TIME) }
                            }
                        }
                    }.show(context, view)
                }
            }

            item {
                labelRes = R.string.sortControversial
                hasNestedItems = true
                callback = {
                    popupMenu {
                        style = R.style.Widget_MPM_Menu_Dark_CustomBackground
                        fixedContentWidthInPx = submenuSize

                        section {
                            title = context.getString(R.string.sortControversial)

                            item {
                                labelRes = R.string.sortNow
                                callback = { postsFragment.controversial(PostTimeSort.HOUR) }
                            }
                            item {
                                labelRes = R.string.sortToday
                                callback = { postsFragment.controversial(PostTimeSort.DAY) }
                            }
                            item {
                                labelRes = R.string.sortWeek
                                callback = { postsFragment.controversial(PostTimeSort.WEEK) }
                            }
                            item {
                                labelRes = R.string.sortMonth
                                callback = { postsFragment.controversial(PostTimeSort.MONTH) }
                            }
                            item {
                                labelRes = R.string.sortYear
                                callback = { postsFragment.controversial(PostTimeSort.YEAR) }
                            }
                            item {
                                labelRes = R.string.sortAllTime
                                callback = { postsFragment.controversial(PostTimeSort.ALL_TIME) }
                            }
                        }
                    }.show(context, view)
                }
            }
        }

    }.show(context, view)
}